macros/src/main/scala/shine/macros/Primitive.scala

```diff
@@ -0,0 +1,416 @@
+          c.Expr(expPrimitivesFromClassDef(cdef))
+        case (cdef: ClassDef) :: (md: ModuleDef) :: Nil =>
+          c.Expr(q"{${expPrimitivesFromClassDef(cdef)}; $md}")
+        case _ => c.abort(c.enclosingPosition, "expected a class definition")
+      }
+    }
+
+    def accPrimitive(annottees: c.Expr[Any]*): c.Expr[Any] = {
+      annottees.map(_.tree) match {
+        case (cdef: ClassDef) :: Nil =>
+          c.Expr(accPrimitivesFromClassDef(cdef))
+        case (cdef: ClassDef) :: (md: ModuleDef) :: Nil =>
+          c.Expr(q"{${accPrimitivesFromClassDef(cdef)}; $md}")
+        case _ => c.abort(c.enclosingPosition, "expected a class definition")
+      }
+    }
+
+    def comPrimitive(annottees: c.Expr[Any]*): c.Expr[Any] = {
+      annottees.map(_.tree) match {
+        case (cdef: ClassDef) :: Nil =>
+          c.Expr(comPrimitivesFromClassDef(cdef))
+        case (cdef: ClassDef) :: (md: ModuleDef) :: Nil =>
+          c.Expr(q"{${comPrimitivesFromClassDef(cdef)}; $md}")
+        case _ => c.abort(c.enclosingPosition, "expected a class definition")
+      }
+    }
+
+    def makeLowerCaseName(s: String): String =
+      s"${Character.toLowerCase(s.charAt(0))}${s.substring(1)}"
+
+    def makeVisitAndRebuild(name: TypeName,
+                            additionalParams: List[ValDef],
+                            params: List[ValDef]): Tree = {
+      val v = q"v"
+      q"""
+        override def visitAndRebuild(
+          $v: shine.DPIA.Phrases.VisitAndRebuild.Visitor): $name
+        = new ${Apply( additionalParams match {
+        case List() => Ident(name)
+        case _ => Apply(Ident(name), additionalParams.map {
+          case ValDef(_, name, _, _) => q"$name"
+        })
+      }, params.map {
+        case ValDef(_, name, tpt, _) => tpt match {
+          case Ident(TypeName("DataType")) | Ident(TypeName("ScalarType")) |
+               Ident(TypeName("BasicType"))     => q"$v.data($name)"
+          case Ident(TypeName("Nat"))           => q"$v.nat($name)"
+          case Ident(TypeName("NatIdentifier")) => q"$v.nat($name)"
+          case Ident(TypeName("NatToNat"))      => q"$v.natToNat($name)"
+          case Ident(TypeName("NatToData"))     => q"$v.natToData($name)"
+          case Ident(TypeName("AccessType"))    => q"$v.access($name)"
+          case Ident(TypeName("AddressSpace"))  => q"$v.addressSpace($name)"
+          // Phrase[ExpType]
+          case AppliedTypeTree((Ident(TypeName("Phrase")), _)) =>
+            q"shine.DPIA.Phrases.VisitAndRebuild($name, $v)"
+          // Vector[Phrase[ExpType]]
+          case AppliedTypeTree((Ident(TypeName("Vector")),
+          List(AppliedTypeTree((Ident(TypeName("Phrase")), _)))))
+               |   AppliedTypeTree((Ident(TypeName("Seq")),
+          List(AppliedTypeTree((Ident(TypeName("Phrase")), _)))))
+          =>
+            q"$name.map(shine.DPIA.Phrases.VisitAndRebuild(_, $v))"
+          case _ =>
+            q"$name"
```

Comment:
Could this be reified into normal code? Could we have a function that matches on the kind and returns the appropriate visitor function?